File: app/static/js/formatting.js
/**
 * Global formatting utilities for dates and amounts
 * 
 * Used across all pages in the application for consistent formatting.
 */

/**
 * Format date from ISO format (YYYY-MM-DD) or Date object to DD/MM/YYYY
 * @param {string|Date} dateStr - Date string in ISO format or Date object
 * @returns {string} Formatted date string in DD/MM/YYYY format, or empty string if invalid
 */
function formatDateDDMMYYYY(dateStr) {
    if (!dateStr) return '';
    
    try {
        let date;
        if (typeof dateStr === 'string') {
            date = new Date(dateStr + 'T00:00:00');
        } else {
            date = dateStr;
        }
        
        if (isNaN(date.getTime())) return '';
        
        const day = String(date.getDate()).padStart(2, '0');
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const year = date.getFullYear();
        
        return day + '/' + month + '/' + year;
    } catch (e) {
        return '';
    }
}

/**
 * Convert date from DD/MM/YYYY to YYYY-MM-DD (for HTML date input)
 * @param {string} dateStr - Date string in DD/MM/YYYY format or ISO format
 * @returns {string} Formatted date string in YYYY-MM-DD format, or empty string if invalid
 */
function convertToDateInputFormat(dateStr) {
    if (!dateStr) return '';
    
    try {
        let date;
        
        // Check if already in YYYY-MM-DD format
        if (/^\d{4}-\d{2}-\d{2}$/.test(dateStr)) {
            return dateStr;
        }
        
        // Try ISO format first
        date = new Date(dateStr + 'T00:00:00');
        if (!isNaN(date.getTime())) {
            const year = date.getFullYear();
            const month = String(date.getMonth() + 1).padStart(2, '0');
            const day = String(date.getDate()).padStart(2, '0');
            return year + '-' + month + '-' + day;
        }
        
        return '';
    } catch (e) {
        return '';
    }
}

/**
 * Format amount to Indian rupee display format (x,xx,xx,xxx)
 * 
 * Examples:
 *   100 -> "100.00"
 *   1000 -> "1,000.00"
 *   100000 -> "1,00,000.00"
 *   1000000 -> "10,00,000.00" (10 lakhs)
 *   10000000 -> "1,00,00,000.00" (1 crore)
 * 
 * @param {number|string} amount - Amount to format
 * @returns {string} Formatted amount string in Indian rupee format
 */
function formatIndianRupee(amount) {
    if (amount === null || amount === undefined || isNaN(amount)) {
        return '0.00';
    }
    
    try {
        const num = Math.abs(Number(amount)).toFixed(2);
        const parts = num.split('.');
        let intPart = parts[0];
        const decPart = parts[1];
        
        if (intPart.length <= 3) {
            return intPart + '.' + decPart;
        }
        
        // Apply Indian numbering format (x,xx,xx,xxx)
        const lastThree = intPart.substring(intPart.length - 3);
        let remaining = intPart.substring(0, intPart.length - 3);
        
        let result = '';
        for (let i = 0; i < remaining.length; i++) {
            if (i > 0 && (remaining.length - i) % 2 === 0) {
                result += ',';
            }
            result += remaining[i];
        }
        
        return result + ',' + lastThree + '.' + decPart;
    } catch (e) {
        return '0.00';
    }
}

/**
 * Format currency with Indian rupee symbol
 * @param {number|string} amount - Amount to format
 * @returns {string} Formatted amount string with ₹ symbol
 */
function formatCurrencyINR(amount) {
    return '₹' + formatIndianRupee(amount);
}

File: app/utils/formatting.py
"""
Centralized formatting utilities for dates and amounts.

These functions ensure consistent formatting across the entire application.
"""

from datetime import date
from typing import Optional, Union


CURRENCY_SYMBOLS = {
    "INR": "₹",
    "USD": "$",
    "EUR": "€",
}


def format_date_dd_mm_yyyy(value: Optional[Union[date, str]]) -> str:
    """
    Format date to DD/MM/YYYY format.
    
    Args:
        value: date object or ISO format string (YYYY-MM-DD)
    
    Returns:
        Formatted date string in DD/MM/YYYY format, or empty string if None
    """
    if value is None:
        return ""
    
    if isinstance(value, str):
        try:
            value = date.fromisoformat(value)
        except (ValueError, TypeError):
            return ""
    
    if isinstance(value, date):
        return value.strftime("%d/%m/%Y")
    
    return ""


def format_indian_rupee(value: Optional[Union[float, int, str]]) -> str:
    """
    Format amount to Indian rupee display format (x,xx,xx,xxx).
    
    Examples:
        100 -> "100.00"
        1000 -> "1,000.00"
        100000 -> "1,00,000.00"
        1000000 -> "10,00,000.00" (10 lakhs)
        10000000 -> "1,00,00,000.00" (1 crore)
    
    Args:
        value: Numeric value to format, can be float, int, or string
    
    Returns:
        Formatted string with Indian rupee convention, or "0.00" if invalid
    """
    if value is None or value == "":
        return "0.00"
    
    # Remove common currency formatting characters before numeric parsing.
    value_str = str(value).replace("₹", "").replace(",", "").strip()
    
    try:
        num = float(value_str)
    except (ValueError, TypeError):
        return "0.00"
    
    # Format to 2 decimal places
    formatted = f"{abs(num):.2f}"
    parts = formatted.split(".")
    int_part = parts[0]
    dec_part = parts[1]
    
    # Apply Indian numbering format (x,xx,xx,xxx)
    if len(int_part) <= 3:
        return int_part + "." + dec_part
    
    # Split into groups of 2 from right, with last group of 3
    last_three = int_part[-3:]
    remaining = int_part[:-3]
    
    # Add commas between groups of 2
    result = ""
    for i, digit in enumerate(reversed(remaining)):
        if i > 0 and i % 2 == 0:
            result = "," + result
        result = digit + result
    
    return result + "," + last_three + "." + dec_part


def format_currency(
    value: Optional[Union[float, int, str]],
    currency: str = "INR",
    decimals: int = 2,
) -> str:
    """Format a numeric value as currency, defaulting to INR formatting."""
    currency_code = (currency or "INR").upper()
    symbol = CURRENCY_SYMBOLS.get(currency_code, f"{currency_code} ")
    safe_decimals = max(0, int(decimals))

    if value is None or value == "":
        if safe_decimals == 0:
            return f"{symbol}0"
        return f"{symbol}0.{('0' * safe_decimals)}"

    value_str = (
        str(value)
        .replace("₹", "")
        .replace("$", "")
        .replace("€", "")
        .replace(",", "")
        .strip()
    )

    try:
        num = float(value_str)
    except (ValueError, TypeError):
        if safe_decimals == 0:
            return f"{symbol}0"
        return f"{symbol}0.{('0' * safe_decimals)}"

    sign = "-" if num < 0 else ""

    if currency_code == "INR":
        formatted = f"{abs(num):.{safe_decimals}f}"
        if safe_decimals > 0:
            int_part, dec_part = formatted.split(".")
        else:
            int_part, dec_part = formatted, ""

        if len(int_part) > 3:
            last_three = int_part[-3:]
            remaining = int_part[:-3]
            grouped = ""
            for i, digit in enumerate(reversed(remaining)):
                if i > 0 and i % 2 == 0:
                    grouped = "," + grouped
                grouped = digit + grouped
            int_part = grouped + "," + last_three

        amount = f"{int_part}.{dec_part}" if safe_decimals > 0 else int_part
        return f"{sign}{symbol}{amount}"

    amount = f"{abs(num):,.{safe_decimals}f}" if safe_decimals > 0 else f"{abs(num):,.0f}"
    return f"{sign}{symbol}{amount}"
